### PR TITLE
Add exponential backoff to topic subscriptions calls

### PR DESCRIPTION
**Pull request by yaron2**
Closes https://github.com/dapr/dapr/issues/3839.

This PR adds exponential retries to the methods that get topic subscriptions from the app using HTTP and gRPC and adds test cases.

### CODE REVIEW

File: pkg/runtime/pubsub/subscriptions.go [hunk 1]
```diff
@@ -8,8 +8,11 @@ import (
 	"path/filepath"
 	"strings"
 
+	"github.com/cenkalti/backoff"
```
Comment (halspang) [inline pkg/runtime/pubsub/subscriptions.go:11]: Should this be using https://github.com/dapr/kit/tree/main/retry?

**Review by halspang — COMMENTED** [review]
Looks good to me overall, had a quick question about consistency in regards to `dapr/kit`.

Also, do we think this would fall under the umbrella of resiliency? Or do we think that stuff in the init path is kind of outside of resiliency for now as it's focusing a lot on invocations/component interactions?

**Comment by yaron2** [comment]
> Looks good to me overall, had a quick question about consistency in regards to `dapr/kit`.
> 
> Also, do we think this would fall under the umbrella of resiliency? Or do we think that stuff in the init path is kind of outside of resiliency for now as it's focusing a lot on invocations/component interactions?

This is not related to the resiliency feature, which focuses on optional, opt-in resiliency policies for the Dapr APIs and are very tight with underlying components behavior and set up. Getting topic configurations, like operator and placement retries, is a runtime init `daprd` concern.

File: pkg/runtime/pubsub/subscriptions.go (hunk 1 above)
Comment (yaron2) [reply pkg/runtime/pubsub/subscriptions.go]: Thanks for the pointer, switched to `kit/retry`.

File: tests/apps/service_invocation_grpc_proxy_server/service.yaml [hunk 2]
```diff
@@ -20,7 +20,6 @@ spec:
         dapr.io/enabled: "true"
         dapr.io/app-id: "grpcproxyserver"
         dapr.io/app-port: "50051"
-        dapr.io/app-protocol: "grpc"
```
Comment (artursouza) [inline tests/apps/service_invocation_grpc_proxy_server/service.yaml:23]: Why reverting this? Is it a left over from debugging?

Comment (yaron2) [reply tests/apps/service_invocation_grpc_proxy_server/service.yaml:23]: Redundant, was never needed.

**Review by artursouza — APPROVED** [review]

**Review by halspang — APPROVED** [review]

**Pull request merged**